src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/HtmlCSharp/DefaultLSPProgressListener.cs

```diff
@@ -0,0 +1,210 @@
+            out Task onCompleted)
+        {
+            var onCompletedSource = new TaskCompletionSource<bool>();
+            var request = new ProgressRequest(
+                onProgressNotifyAsync,
+                timeoutAfterLastNotify,
+                handlerCancellationToken,
+                onCompletedSource);
+
+            if (!_activeRequests.TryAdd(token, request))
+            {
+                onCompleted = null;
+                return false;
+            }
+
+            CompleteAfterDelay(token, request);
+            onCompleted = onCompletedSource.Task;
+            return true;
+        }
+
+        private Task ClientNotifyAsyncListenerAsync(object sender, LanguageClientNotifyEventArgs args)
+            => ProcessProgressNotificationAsync(args.MethodName, args.ParameterToken);
+
+        private async Task ProcessProgressNotificationAsync(string methodName, JToken parameterToken)
+        {
+            if (methodName != Methods.ProgressNotificationName ||
+               !parameterToken.HasValues ||
+               parameterToken[ProgressNotificationValueName] is null ||
+               parameterToken[Methods.ProgressNotificationTokenName] is null)
+            {
+                return;
+            }
+
+            var token = parameterToken[Methods.ProgressNotificationTokenName].ToObject<string>(); // IProgress<object>>();
+
+            if (string.IsNullOrEmpty(token) || !_activeRequests.TryGetValue(token, out var request))
+            {
+                return;
+            }
+
+            var value = parameterToken[ProgressNotificationValueName];
+
+            try
+            {
+                request.HandlerCancellationToken.ThrowIfCancellationRequested();
+                await request.OnProgressNotifyAsync(value, request.HandlerCancellationToken).ConfigureAwait(false);
+            }
+            catch (TaskCanceledException)
+            {
+                // Initial handle request has been cancelled
+                // We can deem this ProgressRequest complete
+                ProgressCompleted(token);
+                return;
+            }
+
+            CompleteAfterDelay(token, request);
+        }
+
+        private void CompleteAfterDelay(string token, ProgressRequest request)
+        {
+            CancellationTokenSource linkedCTS;
+
+            lock (request.RequestLock)
+            {
+                CancelAndDisposeToken(request.TimeoutCancellationTokenSource);
+
+                request.TimeoutCancellationTokenSource = new CancellationTokenSource();
+                linkedCTS = CancellationTokenSource.CreateLinkedTokenSource(
+                    request.TimeoutCancellationTokenSource.Token,
+                    request.HandlerCancellationToken);
+            }
+
+            _ = CompleteAfterDelayAsync(token, request.TimeoutAfterLastNotify, linkedCTS); // Fire and forget
+
+
+            async Task CompleteAfterDelayAsync(string token, TimeSpan delay, CancellationTokenSource cts)
```

Comment:
Specifically nested this function as it should only ever by the above `CompleteAfterDelay(string token, ProgressRequest request)`, with a `CancellationTokenSource` which is a [linked cancellation token meeting these requirements](https://github.com/dotnet/aspnetcore-tooling/pull/2119/files#diff-ef97568d908d1a85a8be9a7a91c660c0R111-R113).

If we call it with a different `CancellationTokenSource` we could have unintended consequences as we dispose of the token after its handling: https://github.com/dotnet/aspnetcore-tooling/pull/2119/files#diff-ef97568d908d1a85a8be9a7a91c660c0R133-R137